signature/policy_eval.go

```diff
@@ -253,6 +253,32 @@ func (pc *PolicyContext) GetSignaturesWithAcceptedAuthor(ctx context.Context, pu
 	return res, nil
 }
 
+// IsImagePolicySigned true iff the policy requirement for the image is not insecureAcceptAnything
+func (pc *PolicyContext) IsImagePolicySigned(publicImage types.UnparsedImage) (res bool, finalErr error) {
```

Comment:
`IsRunningImageWithOptions` … “functional options” would probably allow more long-term extensibility, than adding a single option to the function name. (OTOH it requires tedious code to define the functional options.)

> and then we have a IsInsecure() method on the returned policy object? 

I’m not much of a fan of the policy evaluation returning a multi-valued object, with callers making their own policy decisions on top.

What if we flipped this around, and added a `PolicyContext.Require{SignatureValidationToBeEnforced,InsecureAllowAnythingNotToBeUsed}` (whatever we actually want)? That would imply that a single caller with a single `PolicyContext` always has the same opinion about this policy — and that actually seems appropriate. (A caller can always create two separate `PolicyContext` objects.)

Reply:
I just discovered `GetSignaturesWithAcceptedAuthor` - but apparently [nothing uses it](https://github.com/search?q=lang%3Ago+GetSignaturesWithAcceptedAuthor+-is%3Afork&type=code)?

But it seems to be related in that what we want is "evaluate the policy and require signatures".

One thing I *also* want is something like a `string` description for the *successful* case.  If you look at e.g.

```
[root@cosa-devsh ~]# rpm-ostree status
State: idle
Deployments:
● fedora:fedora/x86_64/coreos/testing
                  Version: 38.20230514.2.0 (2023-05-15T17:23:12Z)
                   Commit: d640da041e110e1abaa8a88cf876d68165e02ad9e1294a6ba8babb9d7a32ed39
             GPGSignature: Valid signature by 6A51BBABBA3D5467B6171221809A8D7CEB10B464
[root@cosa-devsh ~]# 
```

The `GPGSignature` thing there is the success output from gpg.

And that's a very useful thing to log for visibility.

Reply:
> That would imply that a single caller with a single PolicyContext always has the same opinion about this policy — and that actually seems appropriate. (A caller can always create two separate PolicyContext objects.

I'm OK with that, yes.

So are we iterating towards two new APIs?

```
func (pc *PolicyContext) DisallowInsecure();
func (pc *PolicyContext) Evaluate(ctx context.Context, publicImage types.UnparsedImage) (string, error);
```

Where `string` contains some sort of human-readable informational message like the GPG example above?

Reply:
For a bit more future extensibility instead of `string` we could return a `PolicyEvaluationSuccess` object that for now just has a `func (e *PolicyEvaluationSuccess) String() string`.  Maybe it could also have a `func (e *PolicyEvaluationSuccess) Type() string`.

Reply:
> I just discovered `GetSignaturesWithAcceptedAuthor` - but apparently [nothing uses it](https://github.com/search?q=lang%3Ago+GetSignaturesWithAcceptedAuthor+-is%3Afork&type=code)?

Yes, that’s a relic of some feature intended in the very first design… I frankly can’t remember what the original purpose was.

---

> One thing I _also_ want is something like a `string` description for the _successful_ case. If you look at e.g.
…
> And that's a very useful thing to log for visibility.

I think that’s really a “debug logging” feature; adding a `logrus.Debug` with that kind of data makes immediate sense to me.

(Going into the debug log would also make it hard to build a client-side policy overlay on top, which is an advantage in my view — the idea of `policy.json` is that all tools enforce the same policy.)

That currently uses the global `logrus` logger; it could be useful to parametrize that per operation. Maybe only after https://tip.golang.org/doc/go1.21#slog lands, though…

Reply:
> I think that’s really a “debug logging” feature

To expand on this a bit:
- During a `podman pull`, all signatures are recorded locally. So the policy check is reproducible and can be investigated from raw data
- Given that signature enforcement can accept any key from a keyring, recording which one matched can make _a bit_ of a difference (the signature contains a truncated version of the key fingerprint, whereas with a matching key we can record the full value)
- I don’t think the key fingerprint really belongs in the UI. Users aren’t realistically going to verify that value in any tool’s output (and even if some did, it’s really already too late). If an image is supposed to be signed by a specific key, set up the policy to only accept that key in the first place! Recording/listing the key can help when testing/debugging key rollover scenarios or the like, but I can’t see a point during normal production operation.

Reply:
> (Going into the debug log would also make it hard to build a client-side policy overlay on top, which is an advantage in my view — the idea of `policy.json` is that all tools enforce the same policy.)

IOW, that `IsRunningImageAllowed` is named like a predicate, and that it returns a bool without any more details, is a _feature_ in my intuition. That could well be a wrong intuition, or a bad match for some use cases, I don't know.

Reply:
> During a podman pull, all signatures are recorded locally. So the policy check is reproducible and can be investigated from raw data

We did the same in ostree, and it caused performance problems to be constantly re-verifying the signature to display this state.

Further, it's possible for the signature policy to change from when an image was pulled.  So the policy check isn't actually quite reproducible.

> I don’t think the key fingerprint really belongs in the UI.

I don't expect users to go to often verify that manually.  But I do think it's reasonable for auditing/compliance tools to check the presence of this string.  It acts as a "proof of work".

Reply:
> It acts as a "proof of work".

BTW, https://github.com/rpm-software-management/dnf5/issues/617 is a parallel in the dnf world, where also the signature verification state is not cached anywhere and tooling doesn't make it very visible.  If dnf did something like rpm-ostree does, `dnf status` would show the set of keys that were actually used to verify packages, and it would have been hopefully obvious that that set was empty...

Reply:
> > During a podman pull, all signatures are recorded locally. So the policy check is reproducible and can be investigated from raw data
> 
> We did the same in ostree, and it caused performance problems to be constantly re-verifying the signature to display this state.

It seems to me that “when a signature is verified / how that is recorded” is a fairly orthogonal discussion.

> > I don’t think the key fingerprint really belongs in the UI.
> 
> I don't expect users to go to often verify that manually. But I do think it's reasonable for auditing/compliance tools to check the presence of this string.

Sure, an audit log makes sense — that’s not a part of the UI; typically it would be reported to a dedicated machine ASAP.

A later “does this system host anything suspicious” check would better be implemented verifying the stored image against the stored signatures, rather than trusting some local “verified on X and it matched fingerprint Y at that time” record.